src/fiber/fiber.ml

```diff
@@ -1,243 +1,455 @@
+      else
+        let i = i - 1 in
+        match arr.(i) with
+        | None -> assert false
+        | Some x -> loop arr i (x :: acc)
+    in
+    fun a -> loop a (Array.length a) []
+
+  let run (type a) (t : a t) ~iter =
+    let module M = struct
+      type value += X of a
+    end in
+    let rec loop : Jobs.t -> a = function
+      | Empty -> iter () |> Nonempty_list.to_list |> Jobs.exec_fills |> loop
+      | Job (ctx, run, x) -> exec ctx run x Empty
+      | Concat (a, b) -> loop2 a b
+    and loop2 a b =
+      match a with
+      | Empty -> loop b
+      | Job (ctx, run, x) -> exec ctx run x b
+      | Concat (a1, a2) -> loop2 a1 (Jobs.concat a2 b)
+    and exec : 'a. context -> ('a -> effect) -> 'a -> Jobs.t -> a =
+     fun ctx k x jobs ->
+      match k x with
+      | exception Toplevel_exception exn -> Exn_with_backtrace.reraise exn
+      | exception exn ->
+        let exn = Exn_with_backtrace.capture exn in
+        exec ctx.on_error.ctx ctx.on_error.run exn jobs
+      | Done (M.X v) -> v
+      | Done _ -> assert false
+      | Unwind (ctx, k, x) -> exec ctx k x jobs
+      | Read_ivar (ivar, k) -> (
+        match ivar.state with
+        | (Empty | Empty_with_readers _) as readers ->
+          ivar.state <- Empty_with_readers (ctx, k, readers);
+          loop jobs
+        | Full x -> exec ctx k x jobs)
+      | Fill_ivar (ivar, x, k) ->
+        let jobs = Jobs.concat jobs (Jobs.fill_ivar ivar x Empty) in
+        exec ctx k () jobs
+      | Suspend (f, k) ->
+        let k = { ctx; run = k } in
+        f k;
+        loop jobs
+      | Restart (suspended, x, k) ->
+        exec ctx k () (Jobs.concat jobs (Job (suspended.ctx, suspended.run, x)))
+      | Get_var (key, k) -> exec ctx k (Univ_map.find ctx.vars key) jobs
+      | Set_var (key, x, f, k) ->
+        let f () = f () (fun x -> Unwind (ctx, k, x)) in
+        let ctx = { ctx with vars = Univ_map.set ctx.vars key x } in
+        exec ctx f () jobs
+      | Unset_var (key, f, k) ->
+        let f () = f () (fun x -> Unwind (ctx, k, x)) in
+        let ctx = { ctx with vars = Univ_map.remove ctx.vars key } in
+        exec ctx f () jobs
+      | With_error_handler (f, on_error, k) ->
+        let f () = f () (fun x -> Unwind (ctx, k, x)) in
+        let on_error =
+          { ctx; run = (fun exn -> on_error exn Nothing.unreachable_code) }
+        in
+        let ctx = { ctx with on_error } in
+        exec ctx f () jobs
+      | Map_reduce_errors (m, on_error, f, k) ->
+        map_reduce_errors ctx m on_error f k jobs
+      | End_of_fiber () ->
+        let (On_release r) = ctx.on_release in
+        deref r jobs
+      | End_of_map_reduce_error_handler on_release -> deref on_release jobs
+      | Never () -> loop jobs
+      | Fork (a, b) ->
+        let (On_release r) = ctx.on_release in
+        r.ref_count <- r.ref_count + 1;
+        exec ctx a () (Jobs.concat (Job (ctx, b, ())) jobs)
+      | Parallel_iter (l, f, k) -> (
+        match l with
```

Comment:
Why do we special case the list traversals but not the other ones?

Reply:
Lists are quite pervasive in OCaml, feels like list traversals should be optimised